### ISSUE TITLE
remeha2 files

### ISSUE THREAD

**Issue by RobertDeserranno** [Nihilissimis/Remeha2MQTT#2]
My Remeha Calenta is now connected to an esp01 with esp-link installed. Now I can use the Remeha software on my laptop through the virtual COM10 port.
But I am interested in connecting the Remeha with my Home Assistant. What should I do to activate these remeha2mqtt files ?
Upload to the esp01? Can I still use the remeha software afterwards?

**Comment by Nihilissimis** [comment]
Yes, you have to flash your ESP with this code using arduino ide in order to use this. 
you cannot simultaneously use ESP-link so the Remeha software would stop working.

**Comment by RobertDeserranno** [comment]
So I made a new sketch in Arduino IDE and pasted the code from your Remeha2MQTT.ino file into it. Afterwards I downloaded the file Remeha2MQTT.h and added it to the sketch. Also my wifi and mqtt data entered in this h file. When verifying and attempting to upload via the Com port, I always get the message that there is no internet connection. (see appendix). What's going wrong.
Thanks for your help
[remehamqtt.txt](https://github.com/Nihilissimis/Remeha2MQTT/files/12161631/remehamqtt.txt)

**Comment by RobertDeserranno** [comment]
I have solved the problem(s), I changed your ino code according the WiFi code of the BasicOTA example in Arduino IDE. Then I was able to flash the sketch to my ESP en after rebooting I was able to upgrading the file over Wifi.
After installing the board on the Remeha boiler I was able to setup a putty connection over Telnet but there was no connection to  the boiler, every second "bad command try again". Then I changed the serial speed to 9600 and now it is working. See attachment. Also in MQTT explorer there is activity, now I have to make some entity in HA I think.
My boiler is in "vacation" mode, I only use it in the winter to warm up the rooms, I have an OEG sun boiler for hot water and this is also connected by mqtt to HA.
[remehamqtt_versieRD.txt](https://github.com/Nihilissimis/Remeha2MQTT/files/12182619/remehamqtt_versieRD.txt)
![mqttRemeha2](https://github.com/Nihilissimis/Remeha2MQTT/assets/128628161/3dec9ee8-3391-4e53-8115-303e9603122d)

**Comment by Nihilissimis** [comment]
Great that you sorted it out. Currently on holiday so will look in on it when I’m back. On 27 Jul 2023, at 04:13, RobertDeserranno ***@***.***> wrote:﻿
I have solved the problem(s), I changed your ino code according the WiFi code of the BasicOTA example in Arduino IDE. Then I was able to flash the sketch to my ESP en after rebooting I was able to upgrading the file over Wifi.
Aftet installing the board on the Remeha boiler I was able to setup an putty connection over Telnet but there was no connection to  the boiler, every second "bad command try again". Then I changed the serial speed to 9600 and now it is working. See attachment. Also in MQTT explorer there is activity, now I have to make some entity in HA I think.
My boiler is in "vacation" mode, I only use it in the winter to warm up the rooms, I have an OEG sun boiler for hot water and this is also connected by mqtt to HA.
remehamqtt_versieRD.txt

**Comment by RobertDeserranno** [comment]
Niels
I am investigating your software code and would also like to read data related to the identification and counters, among other things.
I have made an additional file and can retrieve the identification data, but converting the raw data to readable data is a problem. Where and how can I find the index, length, ... of the data?
I keep getting meaningless numbers next to each field item

**Comment by Nihilissimis** [comment]
Hi Robert,

I’m currently on holiday. Will come back to you once I’m back.

**Comment by Nihilissimis** [comment]
Hi Robert, 

Not sure what your code does very different from mine. The difference between WiFi.status() and WiFi.waitForConnectResult() should not matter much, I assume, but WiFi can be a strange beast. Also not sure what the WifiUDP library is used for. 
This could also be a just time-out issue, maybe I should make that a configurable item. Although both versions seem to wait 5000ms, just in a different way.

Great if you'd like to contribute with getting the identification data, it's been on my to-do list for awhile, but not very high. As an aside, I'd recommend sending this as MQTT 'retained messages' to the server, so you don't have to query this too often.

For the samples I used rjblake's remeha project. Getting this right involved a bit of trial and error, but once done, it was just the grind of getting it all in the struct format that's the main part of the work.
Rjblake has made an [Excel file](https://github.com/rjblake/remeha/blob/master/XML%20Mapping-Remeha%20Calenta-V2.0.xlsx) that contains all the info. I'm not sure anymore if his indexes start at the same point as mine (either 0 or 1), but that should be easy enough to find out from what I have done.

**Comment by RobertDeserranno** [comment]
Niels, I have a few questions about the Arduino code and mqtt.
For the identification items, I receive the correct HEX and DEC value
according to the Excel file, and for certain items this is OK.
However, for example, for the boiler name, I want to receive "Calenta" and
not "67" instead of "C". I can't do this with 'dtostrf() or itoa().
Furthermore, the OTA charge is also difficult, I regularly have to power
off and power on the ESP8266 to be able to use the OTA connection. Any idea
?
greetings Robert


Copy from putty


[DEBUG] Sending 'request getID1 data' command naar boiler
[DEBUG] Data getID1 available for reading
[DEBUG] 74 Bytes getID1 received
[DEBUG] byteVal 10
[DEBUG] Published: Remeha-Calenta/samples/Boiler dF-Code -> 16  (oke)
[DEBUG] byteVal 7
[DEBUG] Published: Remeha-Calenta/samples/Boiler dU-Code -> 7  (oke)
[DEBUG] byteVal 20
[DEBUG] Published: Remeha-Calenta/samples/Boiler Softw Ver. -> 3.20 (is DEC/10and noet HEX/10)
[DEBUG] byteVal 10
[DEBUG] Published: Remeha-Calenta/samples/Boiler Param Ver. -> 1.60 (is DEC/10and noet HEX/10)
[DEBUG] byteVal 5
[DEBUG] Published: Remeha-Calenta/samples/Boiler ParamType -> 5  (oke)
[DEBUG] byteVal 2
[DEBUG] Published: Remeha-Calenta/samples/Next Service Type -> 2 (oke)
[DEBUG] byteVal 31
[DEBUG] Published: Remeha-Calenta/samples/Boiler SNr -> 49 (should be "1")
[DEBUG] byteVal 43
[DEBUG] Published: Remeha-Calenta/samples/Boiler Name -> 67 (should be "C")
[DEBUG] Waiting 10000 ms
[DEBUG] MQTT-client disconnected. Start reconnect
[DEBUG] Connecting to MQTT-server: 192.168.0.185
[DEBUG] Sending 'request data' command naar boiler
[DEBUG] Data available for reading
[DEBUG] 74 Bytes received
[DEBUG] byteVal beb
[DEBUG] Published: Remeha-Calenta/samples/Aanvoer_temperatuur -> 30.51
[byteValID2.txt](https://github.com/Nihilissimis/Remeha2MQTT/files/12561312/byteValID2.txt)

**Comment by Nihilissimis** [comment]
Thanks for this. Would love to so some of your code as well. Messy is fine at this stage ;)

DEC/HEX is just representation. Once divided by 10 it becomes a float anyway. So this works fine. 

Both Serial and the Name are just a string of 16 bytes ASCII encoded values. So the values struct should work the same, just define the length as 16, e.g.: 
{"Boiler_serie_nummer",1,0,39,16},

It would need some adjustments in the 'decode' function as this assumes values to be 2 byte integers. But decoding itself should be quite simple. I think we should look at introducing a 'type' value in the struct and the functions so the function knows what to do with a 16 byte char array.

F.i. something like this works (I don't have the time now to rework this into a struct, but hope you get the gist of it from an example).

// test code
char serial[17] = {49,52,51,49,49,49,48,50,57,57,50,50,48,32,32,32};
char name[17] = {67,97,108,101,110,116,97,32,32,32,32,32,32,32,32,32};

String serialStr = String(serial);
String nameStr = String(name);

Serial.println(serialStr);
Serial.println(nameStr)

**Comment by Nihilissimis** [comment]
Actually this would need a lot more adjustments. The decode() function is defined as int, so there needs to be a new decodeString() function. Also in the publishMQTT() the inputs values are assumed to be int. It'll need some rework here and there.

**Comment by RobertDeserranno** [comment]
Niels
Due to the decode routine, I had already come to this decision.
I am not a professional programmer, I am retired (so I have some time left) with experience in DOS programs and foxpro (Dutch).
2 years ago I replaced the remeha thermostat with a tado system, this works well, but I miss a number of parameters of the boiler. And since I am working on connecting a number of my devices with Home Assistant, I started looking into connecting the remeha as well.
The question is which module do we get the most interesting information from and is the ESP8266 capable of processing multiple modules? Since the identification module contains static data, the counters module might be more interesting?
I'm working with a truncated version on a UNO board to use it later in the real version on the ESP.
Can I count on your knowledge now and then to further complete this project?
[testUNO.txt](https://github.com/Nihilissimis/Remeha2MQTT/files/12575620/testUNO.txt)
[Rem_identif.txt](https://github.com/Nihilissimis/Remeha2MQTT/files/12575621/Rem_identif.txt)
